src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+        if (use_final_probs && !final_costs.empty()) {
+          typename unordered_map<Token *, BaseFloat>::const_iterator iter =
+              final_costs.find(tok);
+          if (iter != final_costs.end())
+            weight = LatticeWeight(iter->second, 0);
+          else
+            weight = LatticeWeight::Zero();
+        }
+        ofst->SetFinal(cur_state, weight);
+        // for sanity check
+        best_cost_in_chunk_ = std::min(
+            best_cost_in_chunk_, tok->tot_cost + weight.Value1() + weight.Value2());
+      }
+    }
+  }
+  // step 1.3 create final_arc for later appending with the next chunk (in
+  // GetLattice)
+  if (create_final_state) {
+    StateId end_state = ofst->AddState(); // final-state for the chunk
+    ofst->SetFinal(end_state, Weight::One());
+
+    state_label_map_.clear();
+    state_label_map_.reserve(std::min((int32)1e5, config_.max_active));
+    for (Token *tok = active_toks_[frame_end].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      // We assign an unique state label for each of the token in the last frame
+      // of this chunk
+      int32 id = state_label_available_idx_++;
+      state_label_map_[tok] = id;
+      // The final weight has been worked out in the previous for loop and
+      // store in the states
+      // Here, we create a specific final state, and move the final costs to
+      // the cost of an arc connecting to the final state
+      KALDI_ASSERT(ofst->Final(cur_state) != Weight::Zero());
+      Weight final_weight = ofst->Final(cur_state);
+      // Use cost_offsets to guide DeterminizeLatticePruned()
+      // later in GetLattice()
+      // For now, we use extra_cost from the decoding stage , which has some
+      // "future information", as
+      // the final weights of this chunk
+      BaseFloat cost_offset = tok->extra_cost;
+      // We record these cost_offset, and after we appending two chunks
+      // we will cancel them out
+      state_label_final_cost_[id] = cost_offset;
+      Arc arc(0, id, Times(final_weight, Weight(0, cost_offset)), end_state);
+      ofst->AddArc(cur_state, arc);
+      ofst->SetFinal(cur_state, Weight::Zero());
+    }
+  }
+  return (ofst->NumStates() > 0);
+}
+
+template <typename FST, typename Token>
+void LatticeIncrementalDecoderTpl<FST, Token>::AppendLatticeChunks(
+    CompactLattice clat, bool not_first_chunk, int32 last_frame_of_chunk,
+    CompactLattice *olat) {
+  using namespace fst;
+  // step 3.1: Appending new chunk to the old one
+  int32 state_offset = olat->NumStates();
+  if (not_first_chunk)
+    state_offset--; // since we do not append initial state in the first chunk
+
+  // A map between state label and the arc position (index)
+  // the incoming states of these arcs are initial states of the chunk
+  // and the olabel of these arcs are the key of this map (state label)
+  // The arc position are obtained from ArcIterator corresponding to the state
```

Comment:
this is a map from what to pair of (what an what)?  You never clearly state this.

Reply:
forget it, I now see it's not a pair.